### DISCUSSION TITLE
How to send db queries from Rust via tauri_plugin_sql?

### DISCUSSION THREAD

**Discussion by danielniccoli** [tauri-apps/tauri#9937] [Q&A] [answered]
I'm using Tauri v2-beta.

I want to put a file that the user selects via a dialog into a database.

The [file system plugin](https://v2.tauri.app/plugin/file-system/) changed and does not allow reading/writing from/to for arbitrary locations.

> Paths accessed with this API must be relative to one of the base directories so if you need access to arbitrary file system paths, you must write such logic on the core layer instead.

In order to read the file, I must pass the path to the selected file to the backend via IPC. I could read the file and send the bytes back to the frontend via IPC, but since the IPC does not support shared memory, this seems rather memory intensive. Especially for large files.

A solution would be to read the file in the Rust backend, and put it into the database from the backend. The problem is, that the plugin has no documented way of sending a query to the database from the backend.

From what I see in [the source](https://github.com/tauri-apps/tauri-plugin-sql/blob/v2/src/plugin.rs), I should be able to access the plugin's state, which should include the database pool. Something along the lines of

```rust
#[tauri::command]
fn store_document(app_handle: tauri::AppHandle, file_path: String) {
    println!("File path: {}", file_path);
    // Access database pool from app state here
}
```

However, I know too little Rust to even know how to start. I would appreciate a little assistance with this.

**Comment by FabianLars** [comment, marked as answer]
> The [file system plugin](https://v2.tauri.app/plugin/file-system/) changed and does not allow reading/writing from/to for arbitrary locations.

It does if the path is allowed on the scope which you can do via the permissions at build time, the `dialog` plugin (it will automatically extend the scope until the next restart, or forever if you use the persisted-scope plugin) or the fs plugin's scope rust apis.

> In order to read the file, I must pass the path to the selected file to the backend via IPC. I could read the file and send the bytes back to the frontend via IPC, but since the IPC does not support shared memory, this seems rather memory intensive. Especially for large files.

Yeah, good call, especially since you'd be transporting the data twice (reading + writing).

> A solution would be to read the file in the Rust backend, and put it into the database from the backend.

What would this actually solve? Just to be clear, the sql js bindings will also send the file data through the IPC.

Anyway, you currently have to use a db crate (like sqlx) yourself and create a new connection until https://github.com/tauri-apps/plugins-workspace/pull/1381 lands.

**Reply by danielniccoli** [reply]
> What would this actually solve? Just to be clear, the sql js bindings will also send the file data through the IPC.

I think you misread what I wrote. When I read the file in the backend (Rust) and then send it to the database in the backend (Rust) it would not use the JS bindings, no? The only thing passed through the IPC would be the file path.

> until https://github.com/tauri-apps/plugins-workspace/pull/1381 lands

I believe that this is exactly what I need! Any idea if and when this PR will be accepted?

**Reply by FabianLars** [reply]
> I think you misread what I wrote. When I read the file in the backend (Rust) and then send it to the database in the backend (Rust) it would not use the JS bindings, no? The only thing passed through the IPC would be the file path.

Correct, but i assumed you would want to get the file to the js side at some point which will then require the IPC. Otherwise i don't see why you'd want to store the file content in a database. Probably too many assumptions on my side then, apologies.

> I believe that this is exactly what I need! Any idea if and when this PR will be accepted?

no. since it's a small pr, maybe this week but as always, no promises.

**Comment by asadsunny14050** [comment]
i believe the plugin is only for sending db commands from the frontend to the database, otherwise one can just use the rust sqlx crate itself to ru n db command from the core layer or backend
